### PR TITLE
vscode-exts/ms-dotnettools-csharp: Init at 1.23.2

### PR DESCRIPTION
**Pull request by jraygauthier**
###### Motivation for this change

Missing vscode extension. Not possible to install simply using `vscode-utils.extensionsFromVscodeMarketplace` as this depends on dowloaded binaries that needs to be patched.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

Tested in isolation using the following expressions:

```bash
$ cd ~/nixpkgs_latest
$ nix-shell -I nixpkgs=. -p 'with import <nixpkgs> {config={allowUnfree = true;};}; vscode-with-extensions.override {vscodeExtensions = with vscode-extensions; [ms-dotnettools.csharp];}'
$ nix-build -I nixpkgs=. -E 'with import <nixpkgs> {config={allowUnfree = true;};}; vscode-with-extensions.override {vscodeExtensions = with vscode-extensions; [ms-dotnettools.csharp];}'
```

### CODE REVIEW

**Comment by jraygauthier** [comment]
This latest version add a patch that fixes debugging a specific test.

At this point, this extension as been quite extensively tested on projects of significant size. It works impressively well.

File: pkgs/misc/vscode-extensions/default.nix [hunk 1]
```diff
@@ -140,6 +140,8 @@ in
     extractNuGet = callPackage ./python/extract-nuget.nix { };
   };
 
+  ms-dotnettools.csharp = callPackage ./ms-dotnettools-csharp {};
```
Comment (SuperSandro2000) [inline pkgs/misc/vscode-extensions/default.nix:143]: Please sort alphabetically.

File: pkgs/misc/vscode-extensions/ms-dotnettools-csharp/default.nix [hunk 2]
```diff
@@ -0,0 +1,142 @@
+{ lib
+, fetchurl
```
Comment (SuperSandro2000) [inline pkgs/misc/vscode-extensions/ms-dotnettools-csharp/default.nix:2]:
```suggestion
{ fetchurl
```

Please don't use stdenv and lib together.

File: pkgs/misc/vscode-extensions/ms-dotnettools-csharp/update-bin-srcs-lib.sh [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+{installPath, binaries, urls: [.url, .fallbackUrl]}}
+EOF
+)
+
+  1>&2 printf "$ cat %q | jq '%s'\n" "$jsonStorePath" "$jqQuery"
+  cat "$jsonStorePath" | jq "$jqQuery"
+
+ # | jq --slurp '. | add'
+
+  # cat ./bin-srcs.json | jq -r 'keys[] as $k | "\($k) \(.[$k] | .urls[0])"'
```
Comment (SuperSandro2000) [inline pkgs/misc/vscode-extensions/ms-dotnettools-csharp/update-bin-srcs-lib.sh:124]:
```suggestion
```

File: pkgs/misc/vscode-extensions/ms-dotnettools-csharp/update-bin-srcs-lib.sh [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+
+  declare zipShaWStorePath
+  zipShaWStorePath="$(prefetchExtensionZip "$publisher" "$name" "$version")"
+
+  declare zipStorePath
+  zipStorePath="$(echo "$zipShaWStorePath" | tail -n1)"
+  1>&2 echo "zipStorePath='$zipStorePath'"
+
+  function rm_tmpdir() {
+    # 1>&2 echo "Removing \`tmpDir='$tmpDir'\`"
```
Comment (SuperSandro2000) [inline pkgs/misc/vscode-extensions/ms-dotnettools-csharp/update-bin-srcs-lib.sh:40]:
```suggestion
```

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]
Can you please fix the merge conflict? Then I can merge it.

**Comment by jraygauthier** [comment]
@SuperSandro2000: Thanks for the review. Re-based on top of latest which fixes the conflict, applied the suggested changes and re-tested. Everything still works fine.

**Comment by SuperSandro2000** [comment]
Result of `nixpkgs-review pr 100181` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>vscode-extensions.ms-dotnettools.csharp</li>
  </ul>
</details>

**Comment by jraygauthier** [comment]
@SuperSandro2000: Is anything wrong with `nixpkgs-review pr 100181`'s output? Running the command locally does not fail.

**Comment by SuperSandro2000** [comment]
> @SuperSandro2000: Is anything wrong with `nixpkgs-review pr 100181`'s output? Running the command locally does not fail.

Nope but I am not sure if this is all packaged correctly and it is a very big diff so I decided to not merge it.

**Comment by rdk31** [comment]
Will it eventually get merged or really there'll be no way of using the c# ext in vscode on nixos?

**Comment by jraygauthier** [comment]
@rdk31: Would like for this to go forward. However, I wasn't told what was missing / wrong.

**Comment by SuperSandro2000** [comment]
@ofborg wval

**Comment by rdk31** [comment]
@SuperSandro2000: Didn't you missclick eval?

**Comment by SuperSandro2000** [comment]
> @SuperSandro2000: Didn't you misstype eval?

Yeah, fixed.

File: pkgs/misc/vscode-extensions/default.nix [hunk 5]
```diff
@@ -141,6 +141,8 @@ let
         };
       };
 
+      ms-dotnettools.csharp = callPackage ./ms-dotnettools-csharp {};
```
Comment (SuperSandro2000) [inline pkgs/misc/vscode-extensions/default.nix:144]:
```suggestion
      ms-dotnettools.csharp = callPackage ./ms-dotnettools-csharp { };
```

File: pkgs/misc/vscode-extensions/ms-dotnettools-csharp/default.nix [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+        arch = builtins.elemAt kSplit 2;
+        platform = builtins.elemAt kSplit 4;
+      in
+      {
+        inherit name arch platform;
+        installPath = v.installPath;
+        binaries = v.binaries;
+        bin-src = fetchurl {
+          urls = v.urls;
+          sha256 = v.sha256;
```
Comment (SuperSandro2000) [inline pkgs/misc/vscode-extensions/ms-dotnettools-csharp/default.nix:33]:
```suggestion
          inherit (v) sha256;
```

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]
I think there are other places where inherit like this can be used but just a suggestion.
The important part is to fix the eval error.

File: pkgs/misc/vscode-extensions/ms-dotnettools-csharp/default.nix [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+      done
+    }
+
+    patchelf_common() {
+      declare elf="''${1?}"
+
+      patchelf_add_icu_as_needed "$elf"
+      patchelf --add-needed "libssl.so" "$elf"
+      patchelf --set-interpreter "$(cat $NIX_CC/nix-support/dynamic-linker)" \
+        --set-rpath "${stdenv.lib.makeLibraryPath [ stdenv.cc.cc openssl.out icu.out ]}:\$ORIGIN" \
```
Comment (SuperSandro2000) [inline pkgs/misc/vscode-extensions/ms-dotnettools-csharp/default.nix:109]:
```suggestion
        --set-rpath "${lib.makeLibraryPath [ stdenv.cc.cc openssl.out icu.out ]}:\$ORIGIN" \
```

File: pkgs/misc/vscode-extensions/ms-dotnettools-csharp/default.nix [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+    patchelf_common "$vsdbg_dir/vsdbg-ui"
+
+    declare razor_dir="$PWD/${razor.installPath}"
+    unzip_to "${razor.bin-src}" "$razor_dir"
+    chmod a+x "$razor_dir/rzls"
+    touch "$razor_dir/install.Lock"
+    patchelf_common "$razor_dir/rzls"
+  '';
+
+  meta = with stdenv.lib; {
```
Comment (SuperSandro2000) [inline pkgs/misc/vscode-extensions/ms-dotnettools-csharp/default.nix:136]:
```suggestion
  meta = with lib; {
```

File: pkgs/misc/vscode-extensions/ms-dotnettools-csharp/default.nix [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+    license = licenses.mit;
```
Comment (SuperSandro2000) [inline pkgs/misc/vscode-extensions/ms-dotnettools-csharp/default.nix:137]: description is missing.

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

File: pkgs/misc/vscode-extensions/ms-dotnettools-csharp/update-bin-srcs-lib.sh [hunk 10]
```diff
@@ -0,0 +1,154 @@
+#!/usr/bin/env bash
```
Comment (SuperSandro2000) [inline pkgs/misc/vscode-extensions/ms-dotnettools-csharp/update-bin-srcs-lib.sh:1]: This should probably use nix-shell.

**Comment by jbalme** [comment]
IMO we need some utility functions to automate some of this across a few packages.

Inside the vsix there is a package.json with a runtimeDependencies key with platform specific runtime dependencies. That is how VSCode finds and downloads runtime dependencies, in this case, OmniSharp, .NET Debugger, and Razor.

An idiomatic solution would be something that lets us do the following:

```
{vscode-utils, omnisharp, dotnet-debugger, razor-language-server}:
vscode-utils.buildVscodeMarketplaceExtension {
  mktplcRef = {
    name = "csharp";
    publisher = "ms-dotnettools";
    version = "1.23.2";
    sha256 = "0ydaiy8jfd1bj50bqiaz5wbl7r6qwmbz9b29bydimq0rdjgapaar";
  };
  runtimeDependencies = {
    "Omnisharp" = omnisharp;
    "Debugger" = dotnet-debugger;
    "Razor" = razor-language-server;
  }
}
```

...and buildVscodeMarketplaceExtension just does the magic of mapping the dependencies in package.json to the supplied runtimeDependencies,

File: pkgs/misc/vscode-extensions/ms-dotnettools-csharp/default.nix (hunk 9 above)
Comment (jraygauthier) [reply pkgs/misc/vscode-extensions/ms-dotnettools-csharp/default.nix:139]: Done

File: pkgs/misc/vscode-extensions/ms-dotnettools-csharp/update-bin-srcs-lib.sh (hunk 10 above)
Comment (jraygauthier) [reply pkgs/misc/vscode-extensions/ms-dotnettools-csharp/update-bin-srcs-lib.sh:1]: nix-shell shebang added to `update-bin-srcs`. `update-bin-srcs-lib.sh` being a bash function lib sourced by `update-bin-srcs`.

**Comment by jraygauthier** [comment]
@SuperSandro2000: Fixes as requested.

@jbalme: Seems like good idea worth working upon to improve the vscode extension framework for nix. IMO: it seems indeed related to this issue but out of scope. It might be worth opening a separate Issue / PR with a reference to this issue (and other) as examples of what would be improved.

**Comment by SuperSandro2000** [comment]
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 100181` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
1 package built:
</summary>
<ul>
<li>vscode-extensions.ms-dotnettools.csharp</li>
</ul>
</details>

The following issues got detected with the above build packages.
Please fix at least the ones listed with your changed packages:

<details>
vscode-extensions.ms-dotnettools.csharp:
<details>


warning: build-tools-in-build-inputs
unzip is a build tool so it likely goes to `nativeBuildInputs`, not `buildInputs`.

Near pkgs/misc/vscode-extensions/vscode-utils.nix:25:5:

```
   |
25 |     buildInputs = [ unzip ] ++ buildInputs;
   |     ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/build-tools-in-build-inputs.md
</details>
</details>

**Comment by jraygauthier** [comment]
@SuperSandro2000: The above `buildInput` issue with automated check is caused by `pkgs/misc/vscode-extensions/vscode-utils.nix` which is not in our current set of files. Should we fix this as part of this PR or leave it for a separate PR?

**Comment by SuperSandro2000** [comment]
> Should we fix this as part of this PR or leave it for a separate PR?

Separate PR please.

**Pull request merged**